src/App.jsx

```diff
@@ -1,45 +1,243 @@
+const token = `Bearer ${import.meta.env.VITE_PAT}`;
 function App() {
   const [todoList, setTodoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isSaving, setIsSaving] = useState(false);
 
-  function addTodo(title) {
-    const newTodo = { id: Date.now(), title: title, isCompleted: false };
-    setTodoList([...todoList, newTodo]);
-  }
+  useEffect(() => {
+    const fetchTodos = async () => {
+      setIsLoading(true);
 
-  function completeTodo(id) {
-    const updatedTodos = todoList.map((todo) => {
-      if (todo.id === id) {
-        return { ...todo, isCompleted: true };
+      const options = {
+        method: 'GET',
+        headers: {
+          Authorization: token,
+        },
+      };
+      try {
+        const resp = await fetch(url, options);
+        if (!resp.ok) {
+          throw new Error('Failed to load todos');
+        }
+        const response = await resp.json();
+        console.log(response);
+        const fetchedRecords = response.records.map((record) => {
+          const todo = {
+            id: record.id,
+            ...record.fields,
+          };
+          if (!todo.isCompleted) {
+            todo.isCompleted = false;
+          }
+          return todo;
+        });
+
+        setTodoList(fetchedRecords);
+      } catch (error) {
+        setErrorMessage(error.message);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  const addTodo = async (newTodo) => {
+    const payload = {
+      records: [
+        {
+          fields: {
+            title: newTodo.title,
+            isCompleted: newTodo.isCompleted,
+          },
+        },
+      ],
+    };
+    const options = {
+      method: 'POST',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+    try {
+      setIsSaving(true);
+      const resp = await fetch(url, options);
+      if (!resp.ok) {
+        throw new Error('Failed posting todos');
+      }
+      const { records } = await resp.json();
+      const savedTodo = {
+        id: records[0].id,
+        ...records[0].fields,
+      };
+
+      if (!records[0].fields.isCompleted) {
+        savedTodo.isCompleted = false;
+      }
+      setTodoList((prevTodos) => [...prevTodos, savedTodo]);
+    } catch (error) {
+      console.log(error);
+      setErrorMessage(error.message);
+    } finally {
+      setIsSaving(false);
+    }
+  };
+
+  const completeTodo = async (todoId) => {
+    const originalTodo = todoList.find((todo) => todo.id === todoId);
+
+    const checkedTodo = {
+      ...originalTodo,
+      isCompleted: !originalTodo.isCompleted,
+    };
+    setTodoList((prevTodos) =>
+      prevTodos.map((todo) => {
+        if (todo.id === checkedTodo.id) {
+          return checkedTodo;
```

Comment:
As far as React is concerned `checkedTodo` is not a copy of `originalTodo` since it was built outside the this parent function. There are a quite a few ways to tackle this but just remember that state setter functions should received a mutated value if it's not the original.